src/unix/futex.c

```diff
@@ -294,3 +294,103 @@ init_futices(process p)
         halt("failed to allocate futex table\n");
 
 }
+
+/* robust mutex handling */
+
+#define FUTEX_OWNER_DIED	0x40000000
+#define FUTEX_KEY_ADDR(x, o)    ((int *)((u8 *)(x) + (o)))
+
+typedef struct robust_list {
+    struct robust_list *next;
+} *robust_list;
+
+typedef struct robust_list_head {
+    robust_list list;
+    long futex_offset;
+    void *list_op_pending;
+} *robust_list_head;
+
+static boolean valid_user_address(process p, void *a, int len)
+{
+    u64 pg = u64_from_pointer(a) & ~PAGEMASK;
+    u64 end = (u64_from_pointer(a) + len) & ~PAGEMASK;
+
+    for (; pg <= end; pg += PAGESIZE) {
```

Comment:
I would suggest making this a more general function like "vmap_validate_range." Also, I think it would be much more efficient to use rangemap_range_find_gaps with a stack closure and a closed over a valid boolean - just set to false if a gap is found.

Reply:
Refactored as you suggested and it looks a lot nicer!